### DISCUSSION TITLE
Orginizing Fabrics by Ansible Folders

### DISCUSSION THREAD

**Discussion by theotherguy2175** [aristanetworks/avd#4706] [Q&A]
Is it possible to orginize fabrics by folders?

Meaning in some fashion have directories based on each fabric. For instance we could build our CORE network in a directory that hold it's own group_vars and hostvars, and the same thing for the DATA_CENTER, and CAMPUS?

**Comment by ClausHolbechArista** [comment]
Sorry for the late reply.

It is a common pattern to build different fabrics as separate Ansible inventories.

An example could be:
```
avd
├── global_vars
│   ├── management             # Our 'global_vars' plugin will parse one layer of folders for YAML files.
│   │   ├── snmp_settings.yml
│   │   ├── local_users.yml
│   │   └── <...>
│   ├── network_services
│   │   ├── network_services_keys.yml  # Contains only the 'network_services_keys' key
│   │   └── <...>
│   └── <...>
├── playbooks
│   ├── build.yml
│   ├── deploy.yml
│   └── <...>
├── DATA_CENTER
│   ├── group_vars
│   │   ├── DATA_CENTER           # Ansible will parse one layer of folders for YAML files.
│   │   │   ├── misc.yml          # Contains 'fabric_name' and other common settings
│   │   │   └── <...>
│   │   ├── NETWORK_SERVICES
│   │   │   ├── tenant_a.yml      # Contains only the 'tenant_a' key (defined under 'network_services_keys')
│   │   │   └── <...>
│   │   ├── DC1
│   │   │   ├── misc.yml          # Contains 'dc_name' and other common settings
│   │   │   └── <...>
│   │   ├── DC1_SUPER_SPINES
│   │   │   ├── super_spine.yml   # Contains only the 'super_spine' and 'type' keys
│   │   │   ├── l3_edge.yml       # Contains only the 'l3_edge' key
│   │   │   └── <...>
│   │   ├── DC1_POD1
│   │   │   ├── misc.yml          # Contains 'pod_name' and other common settings
│   │   │   └── <...>
│   │   ├── DC1_POD1_SPINES
│   │   │   ├── spines.yml        # Contains only the 'spines' and 'type' keys
│   │   │   └── <...>
│   │   ├── DC1_POD1_L3LEAFS
│   │   │   ├── l3leaf.yml        # Contains only the 'l3leaf' and 'type' keys
│   │   │   └── <...>
│   │   ├── DC1_POD1_RACK1
│   │   │   ├── servers.yml       # Contains only connected endpoints for this rack
│   │   │   └── <...>
│   │   └── <...>
│   ├── host_vars
│   │   ├── dc1-pod1-spine1       # Ansible will parse one layer of folders for YAML files.
│   │   │   └── <...>             
│   │   └── <...>             
│   └── inventory.yml
├── CAMPUS
│   ├── group_vars
│   │   ├── CAMPUS                # Ansible will parse one layer of folders for YAML files.
│   │   │   ├── misc.yml          # Contains 'fabric_name' and other common settings
│   │   │   └── <...>
│   │   ├── NETWORK_SERVICES
│   │   │   ├── tenant_a.yml      # Contains only the 'tenant_a' key (defined under 'network_services_keys')
│   │   │   └── <...>
│   │   ├── CAMPUS_CORE
│   │   │   ├── spine.yml         # Contains only the 'spine' and 'type' keys
│   │   │   └── <...>
│   │   ├── BUILDING1
│   │   │   ├── misc.yml          # Contains 'dc_name' and other common settings
│   │   │   └── <...>
│   │   ├── BUILDING1_CORE
│   │   │   ├── l3spine.yml       # Contains only the 'l3spine' and 'type' keys
│   │   │   └── <...>
│   │   ├── BUILDING1_FLOOR1
│   │   │   ├── l2leaf.yml         # Contains only the 'l2leaf' and 'type' keys
│   │   │   ├── network_ports.yml  # Contains only the 'network_ports' key
│   │   │   ├── misc.yml           # Contains 'pod_name' and other common settings
│   │   │   └── <...>
│   │   └── <...>
│   ├── host_vars
│   │   ├── bld1-fl1-sw1           # Ansible will parse one layer of folders for YAML files.
│   │   │   └── <...>             
│   │   └── <...>             
│   └── inventory.yml
├── ansible.cfg
└── Makefile
```

Your Makefile could contain make targets like:
```
.PHONY: build-dc
build-dc: ## Build DC inventory
<tab>ansible-playbook playbooks/build.yml -i DATA_CENTER/inventory.yml
```

All this is just some example I made up for this answer, but we will work on an example for all this as time allows.

**Reply by Jtown0011** [reply]
Hi @ClausHolbechArista, I was able to get the global_vars management settings working but I'm having issues with the global network_services piece. When I perform a build, none of the network services configuration gets added to the device configuration and I'm wondering what I'm doing wrong.

My goal is to try and define different network services (i.e. tenants) across multiple data centers so they are built the exact same way without having to update the configuration in different folders, I can just update the tenant globally and then push out the configuration to all data centers if needed. Not sure if that is a best practice or if I need to isolate tenants on a per data center basis, thoughts?


In my example, I'm using the same folder structure as yours but let me show you what my network_services_keys.yml file and my NETWORK_CONNECTED/tenant_red.yml looks like.



avd/global_vars/network_services/network_services_keys.yml
```
---
network_services_keys:
  - name: Red
```


avd/DC1/NETWORK_SERVICES/Red.yml
```
---
- name: Red
  mac_vrf_vni_base: 10000
  vrfs:
    - name: Red
      vrf_vni: 10
      svis:
        - id: 10
          name: DMZ
          enabled: true
          ip_address_virtual: 10.1.10.1/24

```

**Reply by carlbuchmann** [reply]
@Jtown0011, yes, that is a sensible approach. You can define all network services centrally and leverage our filtering capabilities to only deploy where needed. Review filter key options here: https://avd.arista.com/5.1/roles/eos_designs/docs/input-variables.html#node-type-network-services-configuration.

As for your error, you need to define top level key `Red:`

```yaml
Red:
  - name: Red
    mac_vrf_vni_base: 10000
    vrfs:
      - name: Red
        vrf_vni: 10
        svis:
          - id: 10
            name: DMZ
            enabled: true
            ip_address_virtual: 10.1.10.1/24
```

Let me know if that resolves the issue for you!

**Reply by Jtown0011** [reply]
@carlbuchmann you're the man! 

That worked, I forgot to define the top level key. I will look into the filter key option you mentioned and let you know if I run into any issues with that. I appreciate the help, thanks!

**Reply by Jtown0011** [reply]
I have another question regarding global_vars, group_vars, and host vars. I hope this doesn't get too confusing but I will try my best to explain what I'm trying to achieve, or if it's even possible.

As I understand it the hierarchy is global_vars then group_vars and finally host_vars.

I'm trying to figure out, if possible, to inherit generic configuration from global_vars all while combining it in with data center specific configuration to make one whole configuration component but I haven't figured out a clean way to do so without overriding the configuration from global_vars to group_vars.

Let's say I have a file **_avd/global_vars/monitoring/snmp_server.yml_** used for snmp server configuration that I want deployed across all data centers. 

Now I would like to configure data center specific snmp configuration in **_avd/DC1/group_vars/DC1.yml_**  For example, things like location and hosts that are specific to the data center.

How can I combine the generic global_vars configuration with the group_vars configuration without overriding the configuration?

**Comment by theotherguy2175** [comment]
This is a huge help and on the surface this works great. The only problem I am having is it would be nice to have 1 AVD project for our whole network. And when testing as long as the fabric was built in the respective folders everything works. But for instance if we had a CORE network with P and PE nodes, I'm not sure how devices could be part of the fabric for DATA_CENTER for instance. 

When I tried running commands to include multiple inventories and yaml files I got an error saying AVD can't build separate fabrics. Is it not possible to somehow have P and PE in a CORE folder and then have CE devices in other folders but still use the same fabric?

**Reply by ClausHolbechArista** [reply]
I would like to understand your pain points in more detail. Why do you need the devices to be in the same AVD project / build?

**Comment by theotherguy2175** [comment]
Just from a simplicity standpoint. Even the globar_vars would be nice. Instead of having 4-5 repos, it could be just 1. But as mentioned having them in different folders makes organization easier, but AVD doesn't seem to have the capability to do a EVPN/VXLAN fabric for example for different folders.

**Reply by ClausHolbechArista** [reply]
We are inheriting some limitations of Ansible around this. Ansible can have one folder of group_vars and host_vars and one inventory.yml file, so there is not much we can do around that.
